app/components/markdown/markdown_image.js

```diff
@@ -1,70 +1,206 @@
-import {Image} from 'react-native';
+import React from 'react';
+import {intlShape} from 'react-intl';
+import {
+    Clipboard,
+    Image,
+    Linking,
+    StyleSheet,
+    Text,
+    TouchableHighlight,
+    View
+} from 'react-native';
 
-export default class MarkdownLink extends PureComponent {
+import FormattedText from 'app/components/formatted_text';
+
+import CustomPropTypes from 'app/constants/custom_prop_types';
+import mattermostManaged from 'app/mattermost_managed';
+import {normalizeProtocol} from 'app/utils/url';
+
+const MAX_IMAGE_HEIGHT = 150;
+
+export default class MarkdownImage extends React.Component {
     static propTypes = {
-        src: PropTypes.string.isRequired
+        children: PropTypes.node,
+        linkDestination: PropTypes.string,
+        onLongPress: PropTypes.func,
+        source: PropTypes.string.isRequired,
+        textStyle: CustomPropTypes.Style
+    };
+
+    static contextTypes = {
+        intl: intlShape.isRequired
     };
 
     constructor(props) {
         super(props);
 
         this.state = {
-            width: 10000,
-            maxWidth: 10000,
-            height: 0
+            width: 0,
+            height: 0,
+            maxWidth: Math.MAX_INT,
+            failed: false
         };
+
+        this.mounted = false;
     }
 
     componentWillMount() {
-        Image.getSize(this.props.src, this.handleSizeReceived, this.handleSizeFailed);
+        this.loadImageSize(this.props.source);
+    }
+
+    componentDidMount() {
+        this.mounted = true;
     }
 
     componentWillReceiveProps(nextProps) {
-        if (this.props.src !== nextProps.src) {
-            Image.getSize(nextProps.src, this.handleSizeReceived, this.handleSizeFailed);
+        if (this.props.source !== nextProps.source) {
+            this.setState({
+                width: 0,
+                height: 0,
+                failed: false
+            });
+
+            this.loadImageSize(nextProps.source);
         }
     }
 
+    componentWillUnmount() {
+        this.mounted = false;
+    }
+
+    loadImageSize = (source) => {
+        Image.getSize(source, this.handleSizeReceived, this.handleSizeFailed);
+    };
+
     handleSizeReceived = (width, height) => {
+        if (!this.mounted) {
+            return;
+        }
+
         this.setState({
             width,
             height
         });
     };
 
     handleSizeFailed = () => {
+        if (!this.mounted) {
+            return;
+        }
+
         this.setState({
-            width: 0,
-            height: 0
+            failed: true
         });
-    }
+    };
 
     handleLayout = (event) => {
+        if (!this.mounted) {
+            return;
+        }
+
         this.setState({
             maxWidth: event.nativeEvent.layout.width
         });
-    }
+    };
+
+    handleLinkPress = () => {
+        const url = normalizeProtocol(this.props.linkDestination);
+
+        Linking.canOpenURL(url).then((supported) => {
```

Comment:
is this expected? do we do this on the webapp?

Reply:
This only happens if their `linkDestination` prop is set which only occurs when the image is inside of a link. This is here because we can't do what the web app does when it turns `[![Logo](https://example.com/logo)](https://example.com)` into `<a href="https://example.com"><img src="https://exampl.ecom/logo"/></a>`